tensor_comprehensions/library/layers.yaml

```diff
@@ -0,0 +1,247 @@
+    -> (I_grad, W1_grad) {{
+      I_grad(n, c, h, w) +=! O_grad(n, m, {sh} * h - kh, {sw} * w - kw) * W1(m, c, kh, kw)
+      W1_grad(m, c, kh, kw) +=! O_grad(n, m, {sh} * h - kh, {sw} * w - kw) * I(n, c, h, w)
+    }}
+
+- name: group_convolution
+  lang: |
+    def group_convolution(float(N,G,C,H,W) I, float(G,F,C,KH,KW) W1, float(G,F) B) -> (O)
+    {
+      O(n, g, f, h, w) +=! I(n, g, c, h + kh, w + kw) * W1(g, f, c, kh, kw)
+      O(n, g, f, h, w) = O(n, g, f, h, w) + B(g, f)
+    }
+
+- name: group_convolution_strided
+  lang: |
+    def group_convolution_strided(float(N,G,C,H,W) I, float(G,F,C,KH,KW) W1, float(G,F) B) -> (O)
+    {{
+      O(n, g, f, h, w) +=! I(n, g, c, {sh} * h + kh, {sw} * w + kw) * W1(g, f, c, kh, kw)
+      O(n, g, f, h, w) = O(n, g, f, h, w) + B(g, f)
+    }}
+
+- name: copy2D
+  lang: |
+    def copy2D(float(M, N) I) -> (O) {
+      O(i, j) = I(i, j)
+    }
+
+- name: copy
+  lang: |
+    def copy(float({dimParams}) I) -> (O) {{
+      O({dimIndices}) = I({dimIndices})
+    }}
+
+- name: cosine
+  lang: |
+    def cosine(float(M) I) -> (O) {
+      O(i) = cos(I(i))
+    }
+
+- name: cosine_similarity
+  lang: |
+    def cosine_similarity(float(M, N) I1, float(M, N) I2) -> (O, sumI1, sumI2) {{
+      sumI1(m) +=! I1(m, n) * I1(m, n)
+      sumI2(m) +=! I2(m, n) * I2(m, n)
+      O(m) +=! (I1(m, n) * I2(m, n)) / fmax(rsqrt(sumI1(m)) * sqrt(sumI2(m)), {eps})
+    }}
+
+- name: add
+  lang: |
+    def add(float(N) A, float(N) B) -> (output) {
+      output(i) = A(i) + B(i) + 1
+    }
+
+- name: abs
+  lang: |
+    def abs(float(M, N) A) -> (O1) {
+      O1(m, n) = fabs(A(m, n))
+    }
+
+- name: layernorm
+  lang: |
+    def layernorm(float(T, B, C) I) -> (O, mean, centered, var)
+    {{
+       mean(t, b) +=! I(t, b, c) / C
+       centered(t, b, c) = I(t, b, c) - mean(t, b)
+       var(t, b) +=! centered(t, b, c) * centered(t, b, c)
+       var(t, b) = (var(t, b) + {eps}) / C
+       O(t, b, c) = centered(t, b, c) / rsqrt(var(t, b))
+    }}
+
+- name: batchnorm
+  lang: |
+    def batchnorm(float(N,C,H,W) I, float(C) rMeanIn, float(C) rVarIn)
+    -> (O, rMeanOut, rVarOut, mean, centered, variance, expectedVariance, normalizedOut)
+    {{
+       mean(c) +=! I(nn, c, hh, ww)
+       mean(c)  = mean(c) / (N * H * W)
+       rMeanOut(c) = (1 - {momentum}) * rMeanIn(c) + {momentum} * mean(c)
+       centered(n, c, h, w) = I(n, c, h, w) - rMeanOut(c)
+       variance(n, c, h, w) = centered(n, c, h, w) * centered(n, c, h, w)
+       expectedVariance(c) +=! (variance(n, c, h, w) + {eps}) / (N * H * W)
+       rVarOut(c) = rsqrt(
+         (1 - {momentum}) * rVarIn(c) + {momentum} * expectedVariance(c))
+       O(n, c, h, w) = centered(n, c, h, w) * rVarOut(c)
+       normalizedOut(n, c, h, w) = O(n, c, h, w)
+    }}
+
+- name: small_mobilenet
+  lang: |
+    def small_mobilenet(float(C1, H, W) I, float(C1, KH1, KW1) W1, float(C1) B1, float(C2, C1) W2, float(C2) B2)
```

Comment:
this is one that we only had in unit tests but not in https://github.com/facebookresearch/TensorComprehensions/tree/master/include/tc/library
please consider opening a task so that we have parity between C++ and Python

Reply:
I think the library referenced in above link is actually out-of-date. It has lots of fcrelu, LUT and matmul/convolution. Nothing else. 
Also, the library is not used anywhere except some C2 tests, we might consider cleaning up this library or start using is more properly and hence maintainig it.

Reply:
#74 issue filed for it

Reply:
#resolved